### DISCUSSION TITLE
behind-the-scenes-of-collection-expressions-part-2-exploring-the-generated-code-list-and-fallback-cases/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#424] [Announcements]
# behind-the-scenes-of-collection-expressions-part-2-exploring-the-generated-code-list-and-fallback-cases/

In this post I look at the code generated when you use collection expressions with List<T> or with types that can be used with collection initializers

https://andrewlock.net/behind-the-scenes-of-collection-expressions-part-2-exploring-the-generated-code-list-and-fallback-cases/

**Comment by daveMC1971** [comment]
Just got bitten by the inject (and accidental registration of z__ReadOnlySingleElementList) in an assembly...An explanation of how this works would be really useful! Thanks for the content!

**Comment by ys0d** [comment]
I am curious about why did they choose such an "optimization"...
```
List<int> list = new List<int>();
CollectionsMarshal.SetCount(list, 5);
```
In this way it will require creating of a new underlying array.
Why not just create a list specifying its initial size:
```
List<int> list = new List<int>(5);
```

There must be a reason for that.. but I can't see it.

**Reply by andrewlock** [reply]
Actually, no, if you don't set the count it doesn't create the original array when you do that 🙂

**Reply by ys0d** [reply]
You are right... Its good to discover smth new, still... :)
But IMHO one-liner would be better instead of two, and wouldn't require any internal magic.

**Reply by andrewlock** [reply]
The extra API is actually more performant I believe (https://github.com/dotnet/runtime/issues/55217#issuecomment-1426793741) because it avoids initializing the elements which are going to be immediately overwritten.

Also, setting the capacity in the constructor sets the Capacity for the list, but it doesn't change the Count, so calling AsSpan() would not get you the backing array, it would return an empty span! You have to use the SetCount API either way, unless you're calling Add or AddRange
